File: 백준/Gold/2653. 안정된 집단/안정된 집단.py
n = int(input())
board = [list(map(int, input().split())) for _ in range(n)]
visited = [False] * n
answer = []
possible = True
for i in range(n):
    if visited[i]:
        continue
    visited[i] = True

    tmp = [i+1]
    for j in range(n):
        if i == j or board[i][j] == 1:
            continue
        if visited[j]:
            possible = False
            break
        visited[j] = True
        tmp.append(j+1)

    if len(tmp) < 2:
        possible = False
        break

    for k1 in tmp:
        for k2 in tmp:
            if board[k1-1][k2-1] == 1:
                possible = False
                break
        if not possible:
            break

    if not possible:
        break

    answer.append(tmp)

if possible:
    print(len(answer))
    for i in range(len(answer)):
        print(*answer[i])
else:
    print(0)

File: 백준/Gold/2653. 안정된 집단/안정된 집단.java
import java.io.IOException;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;
import java.util.ArrayList;
import java.util.List;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;
        int n = Integer.parseInt(br.readLine());
        int[][] board = new int[n][n];
        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < n; j++) {
                board[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        boolean[] v = new boolean[n];
        StringBuilder sb = new StringBuilder();
        int cnt = 0;
        boolean possible = true;
        for (int i = 0; i < n; i++) {
            if (v[i]) continue;
            v[i] = true;
            List<Integer> tmp = new ArrayList<>();
            tmp.add(i);

            for (int j = 0; j < n; j++) {
                if (i == j || board[i][j] == 1) continue;
                if (v[j]) {
                    possible = false;
                    break;
                }
                v[j] = true;
                tmp.add(j);
            }

            if (tmp.size() < 2) {
                possible = false;
                break;
            }

            for (int cur1 : tmp) {
                for (int cur2 : tmp) {
                    if (board[cur1][cur2] == 1) {
                        possible = false;
                        break;
                    }
                }
                if (!possible) {
                    break;
                }
            }
            if (possible) {
                cnt++;
                for (int cur : tmp) {
                    sb.append(cur+1).append(' ');
                }
                sb.append('\n');
            }
        }

        if (possible) {
            System.out.println(cnt);
            System.out.print(sb);
        } else {
            System.out.print(0);
        }
    }
}
